### DISCUSSION TITLE
zonebudget for usg model

### DISCUSSION THREAD

**Discussion by hjia1005** [modflowpy/flopy#1447] [General]
I am using bf.CellBudgetFile to read the cbb file from a usg model and do a zonebudget in python. The current zonbudusg.exe I have produces a weird result and I want to cross check.

Most of the functions in CellBudgetFile work as expected for the usg model except get_ts(). There is an index error but I don't have to use this function.

Another issue is, so far I can only get net flow from the binary file, which is fine for RCH, EVT, WEL but not enough for storage, ghb, riv and some other flow terms. For these I need both in and out. Any help will be appreciated.

**Comment by hjia1005** [comment]
Now figure out that the negative terms are OUT and positive terms are IN. Then did my own zone budget in flopy for the usg model and got a consistent result with the budget summary in the model list file. I also used the usgbud2smp.exe from PEST gw data utilities to process the cbb file and the results are the same as what I got in Python. 

Here is the problem when I use zonbudusg.exe: we literally apply 100% of rainfall to a mine void but the result shows it receives 0 recharge throughout the model run. Also if I add the recharge in all zones together (which covers the entire model), the number doesn't match what is reported in the model list file.

When compared my result with the result from zonbudusg.exe (downloaded from USGS), all other terms are the same (storage, ghb, riv, drn, wel) except rch and evt. I realized that rch and evt have a different data structure in cbc file from other boundary package flow terms.  

What I did was, the first array in rch is the nodes that receive recharge (basically the top cells for each column in the usg model, and their global node number are input to the model as IRCH) and the second array in rch are the flow into that node. But why the two arrays are not zipped as other packages such as ghb, riv and wel? Is there something wrong with this processing? If I am correct, then there could be something wrong with zonbudusg.exe...

![image](https://user-images.githubusercontent.com/33449394/179162930-e2219620-93dc-4c00-b7ad-54006fc1334e.png)
![image](https://user-images.githubusercontent.com/33449394/179162990-c18dccca-18b7-4639-8ed9-51781d982c34.png)

Looking forward to someone from the development team clarifying the RCH and EVT packages in usg model.  Also, if we use NRCHOP=3, then IRCH is omitted, then in what order we should put the RECH array into the model? @cnicol-gwlogic would you be able to look at my question?

**Reply by cnicol-gwlogic** [reply]
I don't really follow your question sorry. Mfusg Rch and evt are array-based data sets, whereas ghb etc are list based. That's why they are not stored in flopy in the same way. 

If using an unstructured grid with different number of nodes per layer, you really must use nevtop and nrchop option values of 2 - i.e. you have to specify the node numbers you apply rch and evt to. Nrchop/nevtop values of 3 will not work as you expect. 

I don't understand your screen shots - what is rch? In flopy you can access the irch and rech arrays using flopy.mfusg.MfusgRch.irch and .rech

Have you looked at your rch file and confirmed that it is in the correct format? 

Can't help more than that sorry, as I don't really understand your issue.

**Reply by hjia1005** [reply]
Hi Chris,

Thank you very much for your reply. Sorry I was in a hurry and didn't include the full code. Variables rch and evt in the screenshot are the data extracted from cbb file using the following code:

CBB = bf.CellBudgetFile('model.cbb')
lst_kstpkper=CBB.get_kstpkper()
for kstpkper in lst_kstpkper:
  rch=CBB.get_data(kstpkper=kstpkper,text='RECHARGE')
  ghb=CBB.get_data(kstpkper=kstpkper,text='HEAD DEP')
  ! do the same for other flow terms
  ! then do something to get the zone budget done

Since I extracted the cell by cell flow, I did the zone budget and exported them to csv files. But the results for rch_in and evt_out are very different from what I got from zonbudusg.exe while other flow terms (storage, ghb, riv and wel),  are the same. Of course exactly same zone file was used.

This model has 13 layers, 66120 rnodes and 401091 gnodes. RCH is applied on the top cell for each vertical column, yes we set both NRCHOP and NEVTOP as 2 and specified the node numbers (used gnode in model file and in the sequence of rnode from 1 to 66120), as shown in the first array of rch in the screenshot.

My result is consistent with the budget summary in the model list file. But zonbudusg gives a different result which worries me, especially it reports 0 recharge in one zone where we applied 100% rainfall as recharge. I really want to know which is wrong.

I hope it's clear now. Thanks again for your time.

**Comment by cnicol-gwlogic** [comment]
Re: "if we use NRCHOP=3, then IRCH is omitted, then in what order we should put the RECH array into the model" - if you look at the mfusg recharge package code, you'll see that for unstructured grids, NRCHOP = 3 means you apply recharge only to the top layer (i.e. layer 1). So the rech array is layer 1 nodes long. If layer 1 does not cover the entire model, I wouldn't use this option unless that is what you want to do (no recharge to layers other than layer 1). From memory, NRCHOP = 1 is the same. I always use 2.

Re: getting a recarray dataset from the recharge / evt cbb arrays for doing your zonebudget checks, you could try something along these lines (note irch is one-based, not zero-based; flopy cbb tools don't seem to convert these, but you can do that if you want):
```python 3
import flopy.utils.binaryfile as bf
import numpy as np


cbb = bf.CellBudgetFile('dubv1tr01.cbc')
#print(cbb.list_records())
rec = cbb.get_data(text='RECHARGE')
#print(rec)
print(cbb.get_times())
print(cbb.get_kstpkper())
print('ntimes in cbc: ', len(rec))
print(rec[0][0].shape, rec[0][0].min(), rec[0][0].max(), rec[0][0].dtype)
print(rec[0][1].shape, rec[0][1].min(), rec[0][1].max(), rec[0][1].dtype)

# example recarray of kper1 recharge
kper1_data = np.rec.fromarrays(
    [rec[0][0].flatten(), rec[0][1].flatten()],
    dtype=[('irch','i'),('rech','f8')]
    )
print(kper1_data)
print(kper1_data['rech'])
print(kper1_data['irch'])

# example recarray of all kpers' recharge
totims = [
    np.array(
        [totim] * rec[kperkstp_index][1].shape[1])
    for kperkstp_index,totim in enumerate(cbb.get_times())
    ]
irch = [
    rec[kperkstp_index][0].flatten()
    for kperkstp_index,totim in enumerate(cbb.get_times())
    ]
rech = [
    rec[kperkstp_index][1].flatten()
    for kperkstp_index,totim in enumerate(cbb.get_times())
    ]

kper_data = np.rec.fromarrays(
    [np.concatenate(totims),
     np.concatenate(irch),
     np.concatenate(rech)],
    dtype=[('totim',float), ('irch',int),('rech',float)])

print(kper_data)
print(kper_data['totim'], kper_data['totim'].shape)
print(kper_data['irch'], kper_data['irch'].shape)
print(kper_data['rech'], kper_data['rech'].shape)
```

Re: your zonbudusg.exe differences - do you have any zones == 0? I seem to recall these being ignored (I could be wrong on that though). I doubt there is something randomly wrong like this with zonbudusg.exe. Maybe try a run with all nodes set to zone 1 or somethng and see if you get any differences.

**Reply by hjia1005** [reply]
Thanks for spending time on the weekend to look into my issue and providing the code. I applied your code to my model and can prove that it's the same as what I understood and processed the recharge arrays.  Yes we tested zonbudusg.exe with many zonations such as taking the whole model as one zone, and all came back with the same results - the sum of rch_in and evt_out from zone(s) are not consistent with the budget summary reported in the model list file. I think zonbudusg.exe should have used a different way to interpret the rch and evt arrays in the cbb file from us. Anyway I will use my own way to do the zone budget before we finally have a same result.

The reason why I want to use NRCHOP and NEVTOP as 3 is we are trying to simulate the "natural" recovery process in the fully dewatered open pit. Ideally RCH and EVT should be applied to the top active (wet) cell in each column. The top active (wet) cell in the pit can change with each stress period as we apply 100% of rainfall as recharge into it, that's why we wish there is a dynamic option such as NRCHOP=3 and NEVTOP =3 to make sure RCH and EVT are effectively working during this process. 

My understanding of these options is based on the online document from Visual Modflow since the mfusg_io_v_1_5.pdf didn't give many details.

https://www.waterloohydrogeologic.com/help/vmod-flex/index.html?vm_translationsettings_rch_and_evt.htm
![rch evt](https://user-images.githubusercontent.com/33449394/179430022-26b0b671-21c0-400e-93d2-51ece1cafa72.PNG)

From mfusg_io_v_1_5.pdf
![image](https://user-images.githubusercontent.com/33449394/179430007-1f17f5bf-f3ca-4ca5-8f63-64db2b6c5295.png)

However, due to the current unexplained discrepancy between the zone budget results, I will be very cautious to try these options.

Thanks again for your time and help. I also wish the developer of zonbudusg.exe can get into this discussion to shed a light on these questions :-)

**Reply by cnicol-gwlogic** [reply]
Re: NRCHOP/NEVTOP = 3, if you look in the usg gwf2rch8u1.f code, you see this:
![image](https://user-images.githubusercontent.com/37314969/179433083-ec062799-9e4c-4dbb-b1db-871c50513b1c.png)
Same for evt (gwf2evt8u1.f):
![image](https://user-images.githubusercontent.com/37314969/179433181-f271f74d-eac5-44ef-aba9-d9e373cfae65.png)

I.e. unless layer 1 covers the entire model, you will have areas of zero rch / evt where deeper layers outcrop.

To me, it sounds like you might need a time-varying IRCH/IEVT (and ETSURF/DEPTH) array, following the pit surface as it descends.

...and for post-mining recovery, LAK package, which will leave EVT/RCH being applied to the final pit surface nodes where the LAK is dry. Where it is wet, LAK will simulate evap / rain from/to the waterbody surface area.

**Reply by hjia1005** [reply]
I tried to avoid reading the source code in Fortran but it is worth doing so when having questions. 

Apparently when option==3 the package will only be applied to the number of nodes equivalent to those in layer 1, definitely not a choice for our models where the top layer normally only covers the alluvium or regolith.

We tried to use a time-varying IRCH/IEVT but we have no idea where the pit surface will reach in the recovery process. We will use LAK package in such situations.

It is a very insightful discussion, I appreciate all your efforts and assistance.

**Reply by cnicol-gwlogic** [reply]
fyi @hjia1005, you might be right re: zonbudusg not dealing correctly with NEVTOP/NRCHOP == 2:
![image](https://user-images.githubusercontent.com/37314969/179441147-b79e96b5-2095-453e-b94b-2a2e9edb41db.png)
NREAL is read from the cbb array header, and is the number of entries in the array. IBUFF is the node number array, and BUFFD is  the cbb rate/budget value array.
I have a feeling that this line may be wrong and will give unexpected results:
BUFFD(IBUFF(I))=BUFFD(I)
...unless the node numbers in IBUFF happen to be 1:NREAL (unlikely for a truly unstructured model, especially one in which someone has gone to the trouble of specifying the nodes to/from which RCH/EVT are applied).  As far as I can tell, in cases in which IBUFF(I) (i.e. the node number to/from which EVT/RCH is applied) exceeds the size of BUFFD, fortran just quietly lets that slide. I'm no fortran expert though so could well be wrong. But I have a feeling this is what is causing your discrepancy between zonbudusg and usgbud2smp / your manual python checks.

I think BUFFD would need to be allocated to a size reflective of the maximum node number in the IBUFF array for the above line to work as expected. But that has its own issues as BUFFD is a replica of BUFF, which is read from the CBB (from 1:NREAL). A different solution would be needed.

I suggest you put together a simple test case showing your error (prove that zonbudusg is not picking up rch assigned via NRCHOP=2 correctly) and send to @langevin-usgs / Sorab (noting that this is not really the place for mfusg / zonebudget issues).

...update: actually I'm not sure this is an issue in zonbudusg. I think buffd will be allocated to the full size of the model because storage is the first array read from the cbb (and I believe must be included for zonbudusg to work) - and the storage array is full 3d of entire model. On subsequent cbb entries, like RCH/EVT, buffd gets reset to zeros before being updated again, but I believe remains at its initally full3d allocated size. In this case, BUFFD(IBUFF(I))=BUFFD(I) should work fine.

**Reply by hjia1005** [reply]
@cnicol-gwlogic, I spent some time reading zonbugusg.f and I think that your original "feel" (before the ...update) was kind of right. 

![image](https://user-images.githubusercontent.com/33449394/179633141-9eb6c6b3-65cb-4fbf-8b3a-de7e248ec76c.png)

Above is the row to read the array head and assign them to those variables (KSTP, KPER, TEXT, NREAL, NROW, NL). NREAL is not global, it's updated whenever there is a new record (you can see there are many 'GO TO 100' in zonbudusg.f). Below shows the array head for each stress period by running CBB.list_records( ):

![image](https://user-images.githubusercontent.com/33449394/179633545-0a009e4c-cf70-4bf3-9403-54f3de796ba7.png)

This is read by "READ(INBUD,END=1000,ERR=1000) KSTP,KPER,TEXT,NREAL,NROW,NL" then start to check ITYPE. So as you mentioned in your previous post, BUFFD is only allocated to a size of the number of rch (evt) cells other than the number of full 3D cells and that's where the discrepancy comes from.

**Reply by cnicol-gwlogic** [reply]
This is what I was thinking originally too. But I'm not so sure now, as those lines are inside a loop (passing over each array for each save time). The first entry in all usg models should I believe be "constant head" (despite no constant heads being in the model); this entry is n nodes (in the whole model) long, even for a compact budget form. And then if FLOW JA FACE is saved, BUFF/BUFFD increases in size again (38806 is the size of the recharge array, and 611856 is the size of buff/buffd == njag - which would always be way bigger than needed for recharge/evt):
![image](https://user-images.githubusercontent.com/37314969/179640957-3f4cb0e0-8ed4-494e-acd3-9cfd745dbc47.png)
I.e. what I thought was your issue is not an issue as far as I can see. But if you can put together a simple example proving there is an issue I am sure someone @ usgs would look at it.

**Reply by hjia1005** [reply]
@cnicol-gwlogic I read the code again, now understand and agree with what you just replied. NREAL is the size of full 3d of the entire model.

Now I think the issue could be with the repeated use of array BUFFD. Could it be possible that BUFFD(I)=DZERO reset some of the previously assigned rch/evt to 0? Another temporary array variable should have been used to read the ITYPE==3 arrays and then assign it to BUFFD. What do you think? I have to confirm it is truly an issue and not my mistake before I submit it to USGS :-)

`         DO 270 I=1,NREAL
         IF(I.NE.IBUFF(I)) THEN
            BUFFD(IBUFF(I))=BUFFD(I)
            BUFFD(I)=DZERO
          END I`

**Reply by hjia1005** [reply]
@cnicol-gwlogic I can confirm the above, BUFFD(I)=DZERO is the issue. I manually checked my zones. If most or all of the gnode numbers in a specific zone are smaller than the size of evt arrays (66120 in my case) then the accumulated recharge or evaporation is very small or 0. Zone2 to zone4 in my model are like this and zone 4 got 0. If most or all of the gnode numbers in a specific zone are larger than the size of evt arrays (66120 in my case) they are well preserved because there is no chance for them to be re-assigned to 0 by BUFFD(I)=DZERO, such as zone 5 to zone 7 in my model. Now I can submit it to USGS.

@langevin-usgs where should I submit this issue related to zonbugusg.exe?

**Reply by cnicol-gwlogic** [reply]
I think you are spot on, yep, what you say makes sense. Need another array. 

I think if you ordered your IRCH/IEVT arrays in node-ascending order you might not see this issue. But still, there is an issue there I think.

**Reply by hjia1005** [reply]
It's like a fight between gnode and rnode ~~~

**Comment by christianlangevin** [comment]
I've been following this discussion a bit while on vacation.  The USGS repository for MODFLOW-USG is located [here](https://github.com/MODFLOW-USGS/mfusg).  It sounds like you may have identified a bug in zonbudusg, which we should fix.  If you had a simple flopy script that could set up a MODFLOW-USG simulation and/or if you could provide the MODFLOW-USG and zonbudusg input files as an attachment to a new issue on the MODFLOW-USG repository, we could work on a fix.

**Reply by hjia1005** [reply]
@langevin-usgs Thanks, the issue is submitted in MODFLOW-USG repository.
[https://github.com/MODFLOW-USGS/mfusg/issues/50](url)

**Reply by christianlangevin** [reply]
Okay, we will take a look.  Not sure of the time frame as it's a busy travel season.  Hopefully you have found a workaround in the meantime.

**Reply by djgraf** [reply]
Following up on this, as I've been having the same issue. Was a fix ever implemented? The binaries list [here](https://github.com/MODFLOW-ORG/executables/releases) indicates that zonbudusg may have been updated on 09/30/2025, but the version number (1.5) is the same as the previous release (02/27/2019).

**Reply by djgraf** [reply]
I ended up tracking down the patched source code and compiling it myself. Seems ZONBUDUSG was never recompiled after the patch, so the latest release doesn't include it. I think it would be helpful for folks to update it at some point, so I figured I would let ya'll know.

**Reply by wpbonelli** [reply]
@djgraf thanks for the heads up. Will look into making a new zonbudusg release.

New dates on the release page don't necessarily imply a new build. That is a bit confusing, sorry. Some binaries are now hosted on GitHub as the old download links were flaky, and it's the GitHub release date getting captured. The version number will change if the binary itself is patched.

**Reply by nchen-eki** [reply]
Hello! Just checking to see whether there have been any updates or fixes to the ZBUD executables regarding this issue. Thank you!
